MF-200 - Enable pagination of result sets

dborovcanin opened the pull request:
Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add pagination to clients and channels endpoints

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

mijicd commented on manager/api/logging.go:
```diff
@@ -86,7 +86,7 @@ func (lm *loggingMiddleware) ViewClient(key string, id string) (client manager.C
 }
 
-func (lm *loggingMiddleware) ListClients(key string) (clients []manager.Client, err error) {
+func (lm *loggingMiddleware) ListClients(key string, offset int, limit int) (clients []manager.Client, err error) {
```
Please change signature to `ListClients(key string, offset, limit int)`.

mijicd commented on manager/api/logging.go:
```diff
@@ -151,7 +151,7 @@ func (lm *loggingMiddleware) ViewChannel(key string, id string) (channel manager
 }
 
-func (lm *loggingMiddleware) ListChannels(key string) (channels []manager.Channel, err error) {
+func (lm *loggingMiddleware) ListChannels(key string, offset int, limit int) (channels []manager.Channel, err error) {
```
Same here.

mijicd commented on manager/api/metrics.go:
```diff
@@ -70,13 +70,13 @@ func (ms *metricsMiddleware) ViewClient(key string, id string) (manager.Client,
 }
 
-func (ms *metricsMiddleware) ListClients(key string) ([]manager.Client, error) {
+func (ms *metricsMiddleware) ListClients(key string, offset int, limit int) ([]manager.Client, error) {
```
And here.

mijicd commented on manager/api/metrics.go:
```diff
@@ -115,13 +115,13 @@ func (ms *metricsMiddleware) ViewChannel(key string, id string) (manager.Channel
 }
 
-func (ms *metricsMiddleware) ListChannels(key string) ([]manager.Channel, error) {
+func (ms *metricsMiddleware) ListChannels(key string, offset int, limit int) ([]manager.Channel, error) {
```
And here.

mijicd commented on manager/api/requests.go:
```diff
@@ -110,16 +110,16 @@ func (req viewResourceReq) validate() error {
 	}
 
-	if req.size > 0 && req.offset >= 0 {
+	if req.offset > 0 && req.limit >= 0 {
```
While i agree with renaming `size` to `limit`, the condition was correct. You can start from the beginning (`offset >= 0`), and you must select at least 1 element (`limit > 0`).

mijicd commented on manager/api/transport.go:
```diff
@@ -247,10 +248,19 @@ func decodeView(_ context.Context, r *http.Request) (interface{}, error) {
 }
 
 func decodeList(_ context.Context, r *http.Request) (interface{}, error) {
+	offset, err := strconv.Atoi(r.FormValue("offset"))
```
Please don't use form values. Instead, use the following approach:

```golang
q := r.URL.Query()
q.Get("offset")
q.Get("limit")
```

Check out the relevant godocs:
- https://golang.org/pkg/net/url/#URL.Query
- https://golang.org/pkg/net/url/#Values.Get

mijicd commented on manager/channels.go:
```diff
@@ -24,8 +24,8 @@ type ChannelRepository interface {
-	// All retrieves the channels owned by the specified user.
-	All(string) []Channel
+	// Bulk retrieves the bulk of channels owned by the specified user.
+	Bulk(string, int, int) []Channel
```
Don't rename.

mijicd commented on manager/clients.go:
```diff
@@ -44,8 +44,8 @@ type ClientRepository interface {
-	// All retrieves the clients owned by the specified user.
-	All(string) []Client
+	// Bulk retrieves the bulk of clients owned by the specified user.
+	Bulk(string, int, int) []Client
```
Same here.

mijicd commented on manager/manager.go:
```diff
@@ -89,7 +89,7 @@ func (ms *managerService) ViewClient(key, id string) (Client, error) {
 }
 
-func (ms *managerService) ListClients(key string) ([]Client, error) {
+func (ms *managerService) ListClients(key string, offset int, limit int) ([]Client, error) {
```
See the remark regarding the method signatures above.

mijicd commented on manager/manager.go:
```diff
@@ -156,7 +156,7 @@ func (ms *managerService) ViewChannel(key, id string) (Channel, error) {
 }
 
-func (ms *managerService) ListChannels(key string) ([]Channel, error) {
+func (ms *managerService) ListChannels(key string, offset int, limit int) ([]Channel, error) {
```
Same here.

mijicd commented on manager/manager_test.go:
```diff
@@ -145,7 +145,7 @@ func TestListClients(t *testing.T) {
 
 	for desc, tc := range cases {
-		_, err := svc.ListClients(tc.key)
+		_, err := svc.ListClients(tc.key, 1, 2)
```
Don't hardcode, use parameters and check various combinations.

mijicd commented on manager/manager_test.go:
```diff
@@ -254,7 +254,7 @@ func TestListChannels(t *testing.T) {
 
 	for desc, tc := range cases {
-		_, err := svc.ListChannels(tc.key)
+		_, err := svc.ListChannels(tc.key, 1, 2)
```
Same here.

mijicd commented on manager/mocks/channels.go:
```diff
@@ -59,7 +59,7 @@ func (crm *channelRepositoryMock) One(owner, id string) (manager.Channel, error)
 }
 
-func (crm *channelRepositoryMock) All(owner string) []manager.Channel {
+func (crm *channelRepositoryMock) Bulk(owner string, offset int, limit int) []manager.Channel {
```
See the remarks given about method signature and renaming.

mijicd commented on manager/mocks/clients.go:
```diff
@@ -63,7 +63,7 @@ func (crm *clientRepositoryMock) One(owner, id string) (manager.Client, error) {
 }
 
-func (crm *clientRepositoryMock) All(owner string) []manager.Client {
+func (crm *clientRepositoryMock) Bulk(owner string, offset int, limit int) []manager.Client {
```
Same here.

mijicd commented on manager/postgres/channels.go:
```diff
@@ -56,11 +56,10 @@ func (cr channelRepository) One(owner, id string) (manager.Channel, error) {
 }
 
-func (cr channelRepository) All(owner string) []manager.Channel {
+func (cr channelRepository) Bulk(owner string, offset int, limit int) []manager.Channel {
```
Same here.

mijicd commented on manager/postgres/channels_test.go:
```diff
@@ -113,7 +113,7 @@ func TestMultiChannelRetrieval(t *testing.T) {
 
 	for desc, tc := range cases {
-		n := len(chanRepo.All(tc.owner))
+		n := len(chanRepo.Bulk(tc.owner, 0, 10))
```
See the remark about test parameters given above.

mijicd reviewed (changes requested):
Besides the requested remarks, please rename all `transport_test.go` to `endpoint_test.go` in a separate commit.

dborovcanin commented on manager/channels.go:
```diff
@@ -24,8 +24,8 @@ type ChannelRepository interface {
-	// All retrieves the channels owned by the specified user.
-	All(string) []Channel
+	// Bulk retrieves the bulk of channels owned by the specified user.
+	Bulk(string, int, int) []Channel
```
@mijicd What do you think of changing offset and limit to uint32 or uint64, according to theirs nature?
Also, what do you think of an idea of wrapping pagination result in pagination object. Something like:
```
type Page struct {
	offset uint32
	limit  uint32
	total  uint32
	data   interface{}
}
```

mijicd replied:
Type choice is correct, though I prefer to use less specific variant. Not sure how would this change affect the signatures (repository & manager). In generally, it's not a bad idea, though some clarification is needed.

dborovcanin replied:
I'm not sure yet, but it should not be problem to Manager or Repository. Limit and offset are already passed as params and data interface is, basically, list of... something. So, repository method would return result type of Page. At the moment, I see only `total` questionable since I will call GORM's `Count` method inside repository method call. Not sure how that would affect performance. Response validation should be straightforward, I don't see any problems right now.

mijicd replied:
I'm talking about request validation, we don't do any response validations. Regarding the second point, I don't recommend to avoid `Page` but return a tuple (array + count) as return value. Executing count will affect performance, but it can be handled, and these endpoints are not time critical anyhow.

dborovcanin committed:
Refactor code

Change method signature and rename Bulk methods back to All.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Rename transport_test.go to endpoint_test.go

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix manager tests to support pagination

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Add default offset and limit support

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Update docs

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Update tests to support pagination

- Move maxLimitSize checking to request validation.
- Add tests to support pagination.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

mijicd commented on manager/api/endpoint_test.go:
```diff
@@ -332,7 +332,7 @@ func TestListClients(t *testing.T) {
 			client: cli,
 			method: http.MethodGet,
-			url:    fmt.Sprintf("%s/clients", ts.URL),
+			url:    fmt.Sprintf("%s/clients?offset=1&limit=2", ts.URL),
```
I'd rather add more cases, covering both usage of default and custom values.

mijicd commented on manager/api/requests.go:
```diff
@@ -110,16 +110,16 @@ func (req viewResourceReq) validate() error {
 	}
 
-	if req.size > 0 && req.offset >= 0 {
+	if req.offset >= 0 && req.limit > 0 {
```
Reading the docs pointed me to another issue: `limit` cannot be infinite. Please set the upper bound and ensure the API documentation is synchronized.

P.S.: don't forget the tests.

mijicd commented on manager/api/transport.go:
```diff
@@ -14,6 +15,8 @@ import (
 	"github.com/prometheus/client_golang/prometheus/promhttp"
 )
 
+const MAX_LIMIT_SIZE = 100
```
See the point given above. Also, **never use java conventions for naming go constants**.

mijicd commented on manager/clients.go:
```diff
@@ -44,8 +44,8 @@ type ClientRepository interface {
 
-	// All retrieves the clients owned by the specified user.
-	All(string) []Client
+	// All retrieves the bulk of clients owned by the specified user.
```
...subset of...

mijicd commented on manager/manager_test.go:
```diff
@@ -244,17 +251,24 @@ func TestListChannels(t *testing.T) {
 	}{
-		"list channels":                        {key, nil},
-		"list channels with wrong credentials": {wrong, manager.ErrUnauthorizedAccess},
+		"list channels":                        {key, 5, nil},
```
Please change to `n`.

mijicd commented on manager/postgres/channels_test.go:
```diff
@@ -113,7 +113,7 @@ func TestMultiChannelRetrieval(t *testing.T) {
 
 	for desc, tc := range cases {
-		n := len(chanRepo.All(tc.owner))
+		n := len(chanRepo.All(tc.owner, 0, 10))
```
Parameterize the test case instead of hard coding values.

mijicd commented on manager/postgres/clients_test.go:
```diff
@@ -130,7 +130,7 @@ func TestMultiClientRetrieval(t *testing.T) {
 
 	for desc, tc := range cases {
-		n := len(clientRepo.All(tc.owner))
+		n := len(clientRepo.All(tc.owner, 0, 10))
```
Parameterize the test instead of hard coding values.

mijicd commented on manager/service.go:
```diff
@@ -61,9 +61,9 @@ type Service interface {
-	// ListChannels retrieves data about all clients that belongs to the user
-	// identified by the provided key.
-	ListChannels(string) ([]Channel, error)
+	// ListChannels retrieves data about bulk of channels specified by offset and limit
```
- `bulk of` => `subset of`
- drop `specified by offset and limit`

mijicd reviewed (changes requested):
In general it looks quite good. Once the requested changes are made it'll be merged.

mijicd commented on manager/api/transport.go:
```diff
@@ -247,10 +248,32 @@ func decodeView(_ context.Context, r *http.Request) (interface{}, error) {
 
 func decodeList(_ context.Context, r *http.Request) (interface{}, error) {
+
+	if len(r.URL.RawQuery) == 0 {
```
This code cannot handle the case where either `offset` or `limit` are provided.

mijicd commented on manager/manager_test.go:
```diff
@@ -136,16 +136,25 @@ func TestListClients(t *testing.T) {
 	}{
-		"list clients":                        {key, nil},
-		"list clients with wrong credentials": {wrong, manager.ErrUnauthorizedAccess},
+		"list clients":                        {key, 0, 5, 5, nil},
```
Missing offset/limit tests.

mijicd commented on manager/manager_test.go:
```diff
@@ -244,17 +253,26 @@ func TestListChannels(t *testing.T) {
 	}{
-		"list channels":                        {key, nil},
-		"list channels with wrong credentials": {wrong, manager.ErrUnauthorizedAccess},
+		"list channels":                        {key, 0, 5, 5, nil},
```
Same here.

mijicd commented on manager/postgres/channels_test.go:
```diff
@@ -105,16 +105,19 @@ func TestMultiChannelRetrieval(t *testing.T) {
 	}{
-		"existing owner":     {email, n},
-		"non-existing owner": {wrong, 0},
+		"existing owner":     {email, 0, n, n},
```
Missing invalid limit check and subset smaller than n.

mijicd commented on manager/postgres/clients_test.go:
```diff
@@ -122,16 +122,19 @@ func TestMultiClientRetrieval(t *testing.T) {
 	}{
-		"existing owner":     {email, n},
-		"non-existing owner": {wrong, 0},
+		"existing owner":     {email, 0, n, n},
```
Same here.

mijicd commented on manager/api/requests.go:
```diff
@@ -110,16 +112,20 @@ func (req viewResourceReq) validate() error {
 	}
 
-	if req.size > 0 && req.offset >= 0 {
+	if req.limit > maxLimitSize {
```
Reject the request if this is the case. Once again, **don't forget to update the API documentation**.

dborovcanin committed:
Fix handling query params for pagination

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix empty result set

Return empty results if invalid offset and limit is passed to channel and client repository.
Update tests accordingly.

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Update manager API docs

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Fix response to invalid limit query param

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

dborovcanin committed:
Remove offset and limmit checks in repository methods

Signed-off-by: Dušan Borovčanin <borovcanindusan1@gmail.com>

mijicd reviewed (approved):
There are minor inconsistencies between the docs and the implementations. These will be fixed in the follow up.

mijicd committed:
Merge branch 'master' into MF-200
